### DISCUSSION TITLE
Fate of component's inner HTML from HTML markup

### DISCUSSION THREAD

**Discussion by azangru** [lit/lit#2237] [General]
Suppose we have a `simple-greeting` Lit component, which outputs "Hello ${name}".

Consider the following html markup:

```html
<simple-greeting name="World">
  <p>
     I come from html file.
  </p>
</simple-greeting>
``` 

Currently, if the `simple-greeting` component uses shadow DOM, it will clear the inner HTML that it received from the markup, and will render its own output in its place. The result will be `Hello World`. If, however, the user overrides the `createRenderRoot` method to avoid the creation of shadow DOM, the component won't clear up the inner HTML, and will render its own content alongside the HTML it received from HTML markup. The result will be: `I come from html file. Hello World.`.

This wasn't the case in lit-element 2.4 (which cleared the initial inner HTML regardless of whether it was using light DOM or shadow DOM), but is now.

- Is this the expected behaviour, or is it a regression?
- Doesn't it seem inconsistent that the component behaves differently in these two scenarios.

[Example in Lit playground](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy5qcyIsImNvbnRlbnQiOiJpbXBvcnQge2h0bWwsIGNzcywgTGl0RWxlbWVudH0gZnJvbSAnbGl0JztcblxuZXhwb3J0IGNsYXNzIFNpbXBsZUdyZWV0aW5nIGV4dGVuZHMgTGl0RWxlbWVudCB7XG4gIHN0YXRpYyBzdHlsZXMgPSBjc3NgcCB7IGNvbG9yOiBibHVlIH1gO1xuXG4gIHN0YXRpYyBwcm9wZXJ0aWVzID0ge1xuICAgIG5hbWU6IHt0eXBlOiBTdHJpbmd9LFxuICB9O1xuXG4gIGNvbnN0cnVjdG9yKCkge1xuICAgIHN1cGVyKCk7XG4gICAgdGhpcy5uYW1lID0gJ1NvbWVib2R5JztcbiAgfVxuXG4gIC8vIG5vIHNoYWRvdyBET01cbiAgY3JlYXRlUmVuZGVyUm9vdCgpIHtcbiAgICByZXR1cm4gdGhpcztcbiAgfVxuXG4gIHJlbmRlcigpIHtcbiAgICByZXR1cm4gaHRtbGA8cD5IZWxsbywgJHt0aGlzLm5hbWV9ITwvcD5gO1xuICB9XG59XG5jdXN0b21FbGVtZW50cy5kZWZpbmUoJ3NpbXBsZS1ncmVldGluZycsIFNpbXBsZUdyZWV0aW5nKTtcbiJ9LHsibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPCFET0NUWVBFIGh0bWw-XG48aGVhZD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi9zaW1wbGUtZ3JlZXRpbmcuanNcIj48L3NjcmlwdD5cbjwvaGVhZD5cbjxib2R5PlxuICA8c2ltcGxlLWdyZWV0aW5nIG5hbWU9XCJXb3JsZFwiPlxuICAgPHA-SSBjb21lIGZyb20gaHRtbCBmaWxlPC9wPlxuICA8L3NpbXBsZS1ncmVldGluZz5cbjwvYm9keT5cbiJ9LHsibmFtZSI6InBhY2thZ2UuanNvbiIsImNvbnRlbnQiOiJ7XG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcImxpdFwiOiBcIl4yLjAuMFwiLFxuICAgIFwiQGxpdC9yZWFjdGl2ZS1lbGVtZW50XCI6IFwiXjEuMC4wXCIsXG4gICAgXCJsaXQtZWxlbWVudFwiOiBcIl4zLjAuMFwiLFxuICAgIFwibGl0LWh0bWxcIjogXCJeMi4wLjBcIlxuICB9XG59IiwiaGlkZGVuIjp0cnVlfV0)

**Comment by thescientist13** [comment]
> _the component won't clear up the inner HTML, and will render its own content alongside the HTML it received from HTML markup ... This wasn't the case in lit-element 2.4_

I believe so, per this thread / answer - https://github.com/lit/lit/issues/2189#issuecomment-925890641

**Comment by justinfagnani** [comment]
> if the simple-greeting component uses shadow DOM, it will clear the inner HTML

No, it just renders to the shadow root and leaves the element's children alone.

> Is this the expected behaviour, or is it a regression?

This is expected behavior now. If you want to clear the container you can do so in `createRenderRoot()`.

> Doesn't it seem inconsistent that the component behaves differently in these two scenarios.

What's inconsistent?

**Reply by azangru** [reply]
> No, it just renders to the shadow root and leaves the element's children alone.

Oh, my mistake, sorry. I assumed that since I wasn't seeing the original markup from the html after a lit element got initialised, it meant that the original children were gone. Didn't occur to me to confirm my guess in the dev tools. I can see now that there are both the shadow root and the original children; but that the browser only displays the content of the shadow root. I didn't know that this is how browsers behave normally.

> What's inconsistent?

Well, I assumed that an element that renders into the shadow DOM clears its children (because I couldn't see them on the page), and the component that uses itself as a render root doesn't. This was an incorrect assumption. Still, I was surprised to see the element's original children displayed on the page in one case, but not in the other.

**Comment by justinfagnani** [comment]
Yeah, you need an unnamed `<slot>` inside your template to render any light DOM children.
